### ISSUE TITLE
Non Deterministic documentation errors build errors 

### ISSUE THREAD

**Issue by renefritze** [pymor/pymor#1020]
AFAICT #1016 did not reliably stop the timeout errors happening in docs building when the ANN training fails in the tutorial.

@HenKlei already had an idea how to fix this and I think that a fix for this issue really should go into the release.

**Comment by sdrave** [comment]
Do you have an example for a failed build?

**Comment by renefritze** [comment]
No. The builds that still failed Friday were restarted.

**Comment by renefritze** [comment]
https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/377488 and https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/377447 have chance to fail soon.

**Comment by HenKlei** [comment]
I will again compare the tutorial and the code in the demo. Maybe there is something I changed in the demo and did not carry over to the tutorial.

Further, I will add a seed for the randomly picking validation snapshots in pyMOR.

Another possibility would be to add a seed for the numpy random generator, maybe there is still some randomization using numpy somewhere in PyTorch (although they claim that setting the seed in PyTorch should be sufficient).

**Comment by sdrave** [comment]
> Further, I will add a seed for the randomly picking validation snapshots in pyMOR.

That should not be necessary. pyMOR manages it's own random state that is used by `sample_randomly` and #1016 should take care to reset it.

**Comment by sdrave** [comment]
@renefritze are you sure those were builds that had #1016 in them? I didn't notice anything ..

**Comment by renefritze** [comment]
> @renefritze are you sure those were builds that had #1016 in them? 

I'm sure. 

> I didn't notice anything ..

That's partly due to docker-in-docker jobs autorestarting on _any_ failure instead of runner failure, see #1018 .

**Comment by HenKlei** [comment]
In the demo, we use `restarts=100` in the reduce method call. When running the demo locally on my machine, this is also necessary. For the default `restarts=10`, I get the same error that no sufficient net could be trained. In the tutorial, we still use the default of `restarts=10`.

However, I have another question: Are we sure that the behavior is non deterministic? Maybe the following happens and we just don't notice every time: What if the build of the documentation **always** needs two runs (since adding the neural network stuff)? That would mean that the random state of the second run is working whereas the one of the first run does not (which in some sense is no error, but random behavior of training neural networks). That would mean that increasing the number of restarts could indeed help here.

**Comment by renefritze** [comment]
A potential automatic rerun of a failed docs builder job would a have a random state with no relation to the first run. The gitlab runner starts a new container to execute the build in, again. Most likely even on a different host. 

How much of a runtime cost is the higher restart count? If it's not too much, lets just do it and be on the safe side.

OTOH, since I haven't seen any ANN related failures since merging #1018 yesterday, and there have been at least 12 full builds since then, maybe the issue is gone after all.

**Comment by HenKlei** [comment]
> A potential automatic rerun of a failed docs builder job would a have a random state with no relation to the first run. The gitlab runner starts a new container to execute the build in, again. Most likely even on a different host.

Okay, did not know that.

> How much of a runtime cost is the higher restart count? If it's not too much, lets just do it and be on the safe side.

The procedure stops immediately if a proper neural network is found, but performs at most `restarts` restarts. So the additional amount of computational costs shouldn't be too much. However, we can not really say that we are save when performing more restarts, there is still some uncertainty.

> OTOH, since I haven't seen any ANN related failures since merging #1018 yesterday, and there have been at least 12 full builds since then, maybe the issue is gone after all.

But you're changes in #1018 did not really change the way of building the docs, right?

**Comment by renefritze** [comment]
> But you're changes in #1018 did not really change the way of building the docs, right?

The change reduced the allowable restart reasons: https://github.com/pymor/pymor/pull/1018/files#diff-230a826add32c4291a3519afe43ef745R66

**Comment by renefritze** [comment]
Can you prepare a commit with the discussed changed for me to cherry-pick into #1024, @HenKlei ?
The issue is still happening there: https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/379842

**Comment by HenKlei** [comment]
> Can you prepare a commit with the discussed changed for me to cherry-pick into #1024, @HenKlei ?

You mean increasing the number of restarts?

Maybe I did not completely understand the discussion regarding the randomization, but at the moment, we reset the randomization state at the beginning of the tutorial. Is it correct that when using `sample_randomly` afterwards, we get a deterministic result even though we do not set a seed? Is that the intended behavior?

**Comment by renefritze** [comment]
> > Can you prepare a commit with the discussed changed for me to cherry-pick into #1024, @HenKlei ?
> 
> You mean increasing the number of restarts?

Yes.

> Maybe I did not completely understand the discussion regarding the randomization, but at the moment, we reset the randomization state at the beginning of the tutorial. Is it correct that when using `sample_randomly` afterwards, we get a deterministic result even though we do not set a seed? 

Somewhat. We _do_ need a seed to get a deterministic `sample_randomly`. However, that seed gets defaulted in `pymor.tools.random.default_random_state` which gets called in `sample_randomly`  since we reset `_default_random_state` in each tutorial. 

> Is that the intended behavior?

What I described just now is, yes.

**Comment by HenKlei** [comment]
Okay, thank you very much for the detailed explanation.

So we can be sure that the pyMOR stuff is indeed deterministic. This means that somewhere, although we set a seed for the PyTorch stuff, there remains some sort of randomization in the PyTorch algorithms. Whenever I ran the demos and so on locally, I never saw a non-deterministic behavior, however. Nevertheless, I will increase the number of restarts since on my machine that is required for the demo (which contains almost the same code) to work. Where do you want to cherry-pick those changes to, @renefritze?

**Comment by renefritze** [comment]
I've just pushed https://github.com/pymor/pymor/pull/1028/commits/5093e600ef730742a0a5039660eb210cbafc8907 after realizing the change was way less involved than I imagined :roll_eyes:

**Comment by sdrave** [comment]
> So we can be sure that the pyMOR stuff is indeed deterministic.

'sure' is a very strong word ...

> Whenever I ran the demos and so on locally, I never saw a non-deterministic behavior, however.

How close did you look. For instance, did you compare the weights of the final ANNs?

In any case: I'm not only concerned about determinism but also about stability of the algorithm. How lucky actually are we that the training succeeds. Should one look, for instance, into changing the ANN topology?

**Comment by HenKlei** [comment]
> How close did you look. For instance, did you compare the weights of the final ANNs?

I compared the sequence of validation losses and they were the same. Maybe I should also look into the weights of the ANNs.

> In any case: I'm not only concerned about determinism but also about stability of the algorithm. How lucky actually are we that the training succeeds. Should one look, for instance, into changing the ANN topology?

I totally agree. It looks like the algorithm worked well (with these network architecture) for the examples from the paper, but the results on new problems aren't as good. However, when I run the FEniCS demo locally, each training run produces an error of 1e-4 or smaller, so in general this seems to work.

Maybe I should try to implement another example at least for the tutorial and further improve the algorithm soon. I could also switch to the 1D example since #1017 seems to have fixed the problems there.

**Comment by sdrave** [comment]
> > How close did you look. For instance, did you compare the weights of the final ANNs?
> 
> I compared the sequence of validation losses and they were the same. Maybe I should also look into the weights of the ANNs.

If the losses are exactly the same, then it's probably assume that your local executions are deterministic.

> > In any case: I'm not only concerned about determinism but also about stability of the algorithm. How lucky actually are we that the training succeeds. Should one look, for instance, into changing the ANN topology?
> 
> I totally agree. It looks like the algorithm worked well (with these network architecture) for the examples from the paper, but the results on new problems aren't as good. However, when I run the FEniCS demo locally, each training run produces an error of 1e-4 or smaller, so in general this seems to work.

Ok, so it's not like you had to do a lot of experimenting to get the example running?

> Maybe I should try to implement another example at least for the tutorial and further improve the algorithm soon. I could also switch to the 1D example since #1017 seems to have fixed the problems there.

Don't know. If there are no fundamental issues with the example except CI behaving weird, I would not change it.

@renefritze, @HenKlei, did you notice any pattern in the CI failures. For instance, could it be that they were all on the same type of runner?

PyTorch docs state that

    Completely reproducible results are not guaranteed across PyTorch releases, individual commits 
    or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions,
    even when using identical seeds.

We already saw before, that the runners seem to be 'different enough' to cause diverging behavior.

**Comment by renefritze** [comment]
Didn't look for runner type. It's always the same torch version with cpu backend though.

**Comment by renefritze** [comment]
With the increased restarts I got no doc building failure in #1028 --> moving to next milestone

**Comment by renefritze** [comment]
Haven't seen this popping up since. Closing.
